pennylane/devices/device_api.py

```diff
@@ -0,0 +1,447 @@
+
+        Args:
+            execution_config (ExecutionConfig): A description of the hyperparameters for the desired computation.
+
+        Returns:
+            Bool
+
+        **Example:**
+
+        For example, the current Python device supports adjoint differentiation if the order is ``1`` and
+        the execution occurs with no shots ``shots=None``.
+
+        >>> config = ExecutionConfig(order=1, shots=None)
+        >>> dev.supports_gradient_with_configuration(config)
+        True
+        >>> config = ExecutionConfig(order=1, shots=10)
+        >>> dev.supports_gradient_with_configuration(config)
+        False
+        >>> config = ExecutionConfig(order=2, shots=None)
+        >>> dev.supports_gradient_with_configuration(config)
+        False
+
+        **Backpropagation:**
+
+        This method is also used to validate support for backpropagation derivatives. Backpropagation
+        is only supported if the devices is transparent to the machine learning framework from start to finish.
+
+        >>> config = ExecutionConfig(diff_method="backprop", framework="torch")
+        >>> python_device.supports_gradient_with_configuration(config)
+        True
+        >>> cpp_device.supports_gradient_with_configuration(config)
+        False
+
+        """
+        if getattr(execution_config, "diff_method", None) == "backprop":
+            return False
+        order = getattr(execution_config, "order", 1)
+        return cls.compute_jacobian != QuantumDevice.compute_jacobian if order == 1 else False
+
+    def compute_jacobian(
+        self,
+        qscript: QScript_or_QScriptBatch,
+        execution_config=None,
+    ):
+        """Calculate the jacobian of either a single or a batch of Quantum Scripts.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            Tuple: The jacobian for each trainable parameter
+
+        See also :meth:`~.supports_jacobian_with_configuration` and :meth:`~.execute_and_compute_jacobian`.
+
+        **Execution Config:**
+
+        The execution config has an ``order`` property that describes the order of gradient requested. If the requested order
+        of gradient is not provided, the device should raise a ``NotImplementedError``. The :meth:`~.supports_jacobian_with_configuration`
+        method can pre-validate supported orders.
+
+        **Return Shape:**
+
+        If a batch of quantum scripts is provided, this method should return a tuple with each entry being the gradient of
+        each individual quantum script. If the batch is of length 1, then the return tuple should still be of length 1, not squeezed.
+
+        """
+        raise NotImplementedError
+
+    def execute_and_compute_jacobian(
```

Comment:
We potentially can compute the Hessian with adjoint for example, given the execution config that can have higher order could we refer to derivative and not jacobian?